content/docs/learn/typed-errors/validation.md

```diff
@@ -0,0 +1,312 @@
+---
+id: validation
+title: Validation
+description: Worked out example of validation.
+sidebar_position: 4
+---
+
+# Validation
+
+<!--- TEST_NAME ValidationTest -->
+
+<!--- INCLUDE .*
+import arrow.core.left
+import arrow.core.right
+import arrow.core.Either
+import arrow.core.NonEmptyList
+import arrow.core.recover
+import arrow.core.raise.*
+-->
+
+This tutorial shows a concrete example of using typed errors to implement
+domain validation. In particular, we begin with the following domain:
+
+```kotlin
+data class Author(val name: String)
+data class Book(val title: String, val authors: List<Author>)
+```
+<!--- KNIT example-validation-01.kt -->
+
+over which we want to implement the following rules:
+
+1. The given title should not be empty,
+2. The list of authors should not be empty,
+3. None of the author names should be empty.
+
+We want to accumulate as many error as possible.
+
+## Smart constructors
+
+The `Author` class exposes its constructor, so we cannot prevent from creating
+wrong values. One could introduce a `require` in the constructor, but we prefer
+to use the typed error mechanism instead. The usual pattern in this case is to
+_hide_ the constructor, and provide a _smart constructor_ by adding an `invoke`
+function in the companion.
```

Comment:
```suggestion
to use the typed error mechanism instead. A common pattern in this case is to
_hide_ the constructor, and provide a _smart constructor_ by adding an `invoke`
operator in the companion.
```

Should we also discuss the trade-offs somewhere? It's been a topic that has been discussed many times, and everytime we show this in a workshop we also get questions around this topic.

To summarise (if memory serves me right):

##### `operator fun invoke`
 - Pro: allows you to _mimic_ a real constructor call, **and** allow you to make the `constructor` private.
 - Pro: Doesn't require additional import
 - Con: Click-through on `Author` brings you `companion object`, you need to click `(..)` braces to find definition site

##### named companion function
 - Pro: allow you to make the `constructor` private.
 - Con: requires additional import
 - Con: Doesn't mimic real constructor call
 
#### Top-level `fun Author`
 - Pro: allows you to _mimic_ a real constructor call, **but** doesn't allow you to make the `constructor` private.
 - Con: doesn't allow the `constructor` to be private
 - Con: Requires additional import

cc\\ @clovis-ai

Reply:
Ahah, I love that you cc'd me here. Sorry for being late, but another important consideration for me is how it reads for beginners: if I share the link to a Dokka page of an `operator fun` function in a companion object to someone who just started learning Kotlin, they have no idea how to call it. Both other solutions are obvious (and IDEA will take care of the imports).